macros/Compat.ejs

```diff
@@ -1,66 +1,436 @@
+  chrome: 'Chrome',
+  firefox: 'Firefox',
+  edge: 'Edge',
+  ie: 'Internet Explorer',
+  opera: 'Opera',
+  safari: 'Safari',
+  servo: 'Servo'
+};
+
+const mobileBrowsers = {
+  webview_android: 'Android',
+  chrome_android: s_chrome_android,
+  edge_mobile: 'Edge mobile',
+  firefox_android: s_firefox_android,
+  ie_mobile: 'IE mobile',
+  opera_android: 'Opera Android',
+  safari_ios: 'iOS Safari',
+}
 
-if ($0) {
-  try {
-    var data = mdn.fetchJSONResource(url);
-  } catch (e) {
-    output = 'File "' + file + '" not found or malformed.'
+var notesArray = [];
+
+/*
+Write the table header.
+
+`browserPlatformType` is either "mobile" or "desktop"
+`browserNames` is either the "mobileBrowsers" or "desktopBrowsers" object
+*/
+function writeTableHead(browserPlatformType, browserNames) {
+  let browserNameKeys = Object.keys(browserNames);
+  let output = `<div id="compat-${browserPlatformType}"><table class="compat-table"><thead><tr>`;
+  output +=  `<th>${localize(compatStrings, 'feature')}</th>`
+  for (let browserNameKey of browserNameKeys) {
+    output += `<th>${browserNames[browserNameKey]}</th>`;
   }
+  output += '</tr></thead>';
+  return output;
 }
 
-if (data) {
-    if (leaf !== '') {
-      output = template("CompatDataTable", [data[leaf], "api"]);
+/*
+Given the value of `version_added` or `version_removed`, this returns
+a string to appear in the table cell, like "Yes", "No" or "?"
+
+`versionInfo` is either null, true, false or a string containing a version number
+*/
+function getVersionString(versionInfo) {
+  switch (versionInfo) {
+    case null:
+      return `<span title="${localize(compatStrings, 'supportsShort_unknown_title')}"
+              style="color: rgb(255, 153, 0);">${localize(compatStrings, 'supportsShort_unknown')}</span>`;
+      break;
+    case true:
+      return `<span style="color: #888">(${localize(compatStrings, 'supportsShort_yes')})</span>`;
+      break;
+    case false:
+      return `<span style="color: #f00">(${localize(compatStrings, 'supportsShort_no')})</span>`;
+      break;
+    default:
+      return versionInfo;
+  }
+}
+
+/*
+Generate the note for a browser flag or preference
+First checks version_added and version_removed to create a string indicating when
+a preference setting is present. Then creates a (browser specific) string
+for either a preference flag or a compile flag.
+
+`supportData` is a support_statement
+`browserId` is a compat_block browser ID
+*/
+function writeFlagsNote(supportData, browserId) {
+  let output = '';
+
+  const firefoxPrefs = 'To change preferences in Firefox, visit about:config.';
+  const chromePrefs = 'To change preferences in Chrome, visit chrome://flags.';
+
+  if (typeof(supportData.version_added) === 'string') {
+    output = 'From version ' + supportData.version_added;
+  }
+
+  if (typeof(supportData.version_removed) === 'string') {
+    if (output) {
+      output += ' until version '+ supportData.version_removed + ' (excluded): ';
```

Comment:
Sorry I should have caught this before. Just a nit, but "exclusive" is better English than "excluded" here (and at line 140).